gap/SnAnUnknownDegree.gi

```diff
@@ -0,0 +1,345 @@
+        local
+            # integer, loop variable
+            a,
+            # elements, in G
+            r, tPower, tPowerOld, c;
+        # Steps 2 & 3: New involution
+        # Check if we either tried enough conjugates or constructed enough
+        # three cycle candidates for the current involution t.
+        # If this is the case, we need to construct the next involution
+        if nrTriedConjugates >= C or nrThreeCycleCandidates >= T then
+            r := RandomElm(ri,"simplesocle",true)!.el;
+            a := 0;
+            tPower := r ^ M;
+            # Invariant: tPower = (r ^ M) ^ (2 ^ a)
+            repeat
+                a := a + 1;
+                tPowerOld := tPower;
+                tPower := tPower ^ 2;
+            until a = logInt2N or isone(ri)(tPower);
+            if a = logInt2N then
+                return NeverApplicable;
+            fi;
+            t := tPowerOld;
+            nrInvolutions := nrInvolutions + 1;
+            nrTriedConjugates := 0;
+            nrThreeCycleCandidates := 0;
+        fi;
+        # Steps 4 & 5: new three cycle candidate
+        # Try to construct a three cycle candidate via a conjugate of t. See
+        # the comment above this function.
+        nrTriedConjugates := nrTriedConjugates + 1;
+        c := t ^ RandomElm(ri,"simplesocle",true)!.el;
+        if not isequal(ri)(t * c, c * t) then
+            nrThreeCycleCandidates := nrThreeCycleCandidates + 1;
+            return (t * c) ^ 2;
+        else
+            return fail;
+        fi;
+    end;
+    # construct the iterator
+    oneThreeCycleCandidate := function()
+        local candidate;
+        repeat
+            if nrInvolutions >= B
+                and (nrTriedConjugates >= C or nrThreeCycleCandidates >= T)
+            then
+                # We are done and were not able to recognize Sn or An.
+                return TemporaryFailure;
+            fi;
+            candidate := tryThreeCycleCandidate();
+            if candidate = NeverApplicable then
+                return NeverApplicable;
+            fi;
+        until candidate <> fail;
+        return candidate;
+    end;
+    return oneThreeCycleCandidate;
+end);
+
+# G: the group to recognize
+# c: possibly a 3-cycle
+# returns a list of group elements. If G is isomorphic to an alternating or
+# symmetric group and c is a 3-cycle, then this function returns a list of
+# bolstering elements with respect to c.
+BindGlobal("BolsteringElements",
+function(ri, c, eps, N)
+    local result, R, S, prebolsteringElms, i, r, cr, cr2;
+    result := [];
+    R := Int(Ceil(7 / 4 * Log2(Float(eps ^ -1))));
+    S := 7 * N * R;
+    prebolsteringElms := [];
+    i := 0;
+    # find pre-bolstering elements
+    while i <= S and Length(prebolsteringElms) <= R do
+        r := RandomElm(ri,"simplesocle",true)!.el;
+        # test whether r is pre-bolstering
+        cr := c ^ r;
+        cr2 := c ^ (r ^ 2);
+        if not isone(ri)(Comm(cr, c))
```

Comment:
Computing a commutator is expensive for e.g. matrices where it requires two inversions and three multiplications. For permutations, it still requires one inversion, one conjugation and one multiplication. But we can get away with two multiplications here:
```suggestion
        if not isequal(ri)(cr*c, c*cr)
```
Of course you are not yet tuning for performance, but I thought I could just mention this thoughts while I had them :-).

One might be able to do even better for specific groups; and also, strictly speaking, it isn't necessary to produce temporary objects for the two products, one could write optimized code which doesn't need to produce the full two products as intermediate objects...

So perhaps we should have an `AreCommuting(ri)(g,h)` operation (as part of `ri`, to deal with projectives *sigh*) which does the "best" depending on the inputs?

Reply:
This is a very good idea. We try to add a framework for this operation.